linked_list/linked_list.py

```diff
@@ -1,102 +1,183 @@
+    # Space Complexity: O(n)
     def get_last(self):
-        pass
+        if self.head == None:
+            return None
+
+        current = self.head
+        while current.next != None:
+            current = current.next
+
+        return current.value
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def add_last(self, value):
-        pass
+        new_node = Node(value)
+
+        if self.head == None:
+            self.head = new_node
+            return 
+
+        current = self.head
+        while current.next != None:
+            current = current.next
+        
+        current.next = new_node
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max(self):
-        pass
+        if self.head == None:
+            return None
+
+        max_value = self.head.value
+        current = self.head.next
+
+        while current != None:
+            if max_value < current.value:
+                max_value = current.value
+            current = current.next
+
+        return max_value
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def delete(self, value):
-        pass
+        current = self.head
+        prev = None
+
+        try:
+            while current.value != value:
+                prev = current
+                current = current.next
+        except:
+            return None
+
+        if current == self.head:
+            self.head = current.next
+        else:
+            prev.next = current.next
+
 
     # method to print all the values in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def visit(self):
         helper_list = []
         current = self.head
 
         while current:
             helper_list.append(str(current.value))
             current = current.next
-        
+
         print(", ".join(helper_list))
 
     # method to reverse the singly linked list
     # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def reverse(self):
-        pass
-  
+        prev = None
+        current = self.head
+        next = None
+
+        while current != None:
+            next = current.next
+            current.next = prev
+            prev = current
+            current = next
+        self.head = prev
+
     ## Advanced/ Exercises
     # returns the value at the middle element in the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def find_middle_value(self):

```

Comment:
👍 , but similarly to search the space complexity is O(1)


Good re-use of existing methods